{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BIOM262: ChIP-Seq workshop – Tuesday Feb 4th 2020"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Introduction:**\n",
    "\n",
    "This workshop will walk you through an example of ChIP-seq analysis. We will focus on running tools as command lines and simple bash scripts – I recommend having a cheat-sheet like this [one](http://cheatsheetworld.com/programming/unix-linux-cheat-sheet/). \n",
    "\n",
    "We will use some common tools such as: \n",
    "* **bowtie2** for alignment (<a href=\"http://bowtie-bio.sourceforge.net/bowtie2/index.shtml\" target=\"_blank\">http://bowtie-bio.sourceforge.net/bowtie2/index.shtml</a>),\n",
    "* **IGV** for visualization (<a href=\"http://software.broadinstitute.org/software/igv/home\" target=\"_blank\">http://software.broadinstitute.org/software/igv/home</a>) \n",
    "* Most of the workshop will be done using **HOMER** (e.g., QC of the data, peak calling etc.; <a href=\"http://homer.ucsd.edu/homer/\" target=\"_blank\">http://homer.ucsd.edu/homer/</a>). HOMER was created by Chris Benner at UCSD, and I love the documentation and tutorials and the threaded humor. To install HOMER follow <a href=\"https://github.com/biom262/cmm262-2020/blob/master/Module_5/Notebooks/Install_Homer.ipynb\" target=\"_blank\">these instructions</a>.\n",
    "\n",
    "To load bowtie2 type:\n",
    "```\n",
    "module load bowtie2\n",
    "```\n",
    "During the workshop, and in general, it is always good to type the command and get the notes and use options of the command. Thus e.g., typing bowtie2 would yield this output (capped after several lines):\n",
    "\n",
    "```\n",
    "[ucsd-trainXX@tscc-login1 ~]$  bowtie2\n",
    "No index, query, or output file specified!  \n",
    "Bowtie 2 version 2.3.0 by Ben Langmead (langmea@cs.jhu.edu, www.cs.jhu.edu/~langmea)   \n",
    "Usage: bowtie2 [options]* -x <bt2-idx> {-1 <m1> -2 <m2> | -U <r>} [-S <sam>]\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## Part I\n",
    "We will start with FASTQ files and perform many of the basic analysis tasks that one might normally do when analyzing ChIP-seq data. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **1.** Align FASTQ reads using bowtie2.\n",
    "The fastqs are at: ` /oasis/tscc/scratch/biom200/cmm262/Module_5/fastqs/`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These files are originally from the following study investigating the roles that reprogramming factors play when transforming MEF (fibroblasts) into embryonic stem cells.\n",
    "[Chronis et al. Cooperative Binding of Transcription Factors Orchestrates Reprogramming](https://www.ncbi.nlm.nih.gov/pubmed/28111071)\n",
    "Sequencing Data: [GSE90893](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE90893)\n",
    "\n",
    "For this tutorial we extracted the ChIP-seq experiments for several transcription factors and histone modifications performed on ESC (embryonic stem cells). To reduce runtimes, only reads that mapped to chr17 (and chr17_random) are included. (the random chromosomes are explained in this link – http://genome.ucsc.edu/FAQ/FAQdownloads#download10.)\n",
    "\n",
    "**Question:** how would you generate such a file with only one chromosome?\n",
    "\n",
    "**TIP:** If you start overwriting your own line (which means you can no longer see what you typed earlier) try the following command: \n",
    "```\n",
    "eval $(resize)\n",
    "```\n",
    "\n",
    "**Note:** if you get a memory error, grab an interactive node:\n",
    "\n",
    "```\n",
    "qsub -I -l nodes=1:ppn=8 -l walltime=4:00:00 -q hotel\n",
    "```\n",
    "\n",
    "To align the reads we will use bowtie2. It is always a good practice to look at the manual of each tool you use, so you get an idea of options and documentation.\n",
    "Initially, to get an idea of the command we will run it on one file as follows:\n",
    "\n",
    "\n",
    "```\n",
    "bowtie2 -p 8 -x /oasis/tscc/scratch/biom200/cmm262/Module_5/mm9/mm9_index \\\n",
    "    -U /oasis/tscc/scratch/biom200/cmm262/Module_5/fastqs/oct4-esc.chr17.2m.fastq | \\\n",
    "    samtools view -bS -t /oasis/tscc/scratch/biom200/cmm262/Module_5/mm9/mm9.fa.fai > \\\n",
    "    <path to your bams directory>/oct4-esc.chr17.2m.bam\n",
    "```\n",
    "\n",
    "(notice if you have very long lines of code you can break it up with a \"\\\\\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To do it properly, we will use a for loop to get BAMs from all the fastqs in the directory:\n",
    "```\n",
    "for f in <path to fastqs directory>/*fastq; \n",
    "do fname=`basename $f .fastq`; \n",
    "    bowtie2 -p 8 -x /oasis/tscc/scratch/biom200/cmm262/Module_5/mm9/mm9_index -U $f | \\\n",
    "    samtools view -bS -t /oasis/tscc/scratch/biom200/cmm262/Module_5/mm9/mm9.fa.fai > \\\n",
    "    <path to bams directory>/$fname.bam ; \n",
    "done\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This will produce BAM files for the 6 datasets. HOMER can analyze SAM files and if it receives BAM files it converts them to SAM so samtools has to be available (you can check that by typing “samtools” in the command line). \n",
    "\n",
    "It is a good practice to always double check datasets before you start analyzing them. For instance, use samtools to view the files.\n",
    "\n",
    "``` \n",
    "samtools view -h <path to bams directory>/input-esc.chr17.2m.bam | less\n",
    "```\n",
    "\n",
    "and validate that the files are indeed what they should be (e.g., aligned to chr 17, and have 2M reads). \n",
    "\n",
    "To calculate the number of reads do:\n",
    "\n",
    "```\n",
    "samtools view bams/input-esc.chr17.2m.bam | wc -l\n",
    "```\n",
    "Alternatively, another option that is even better for here is to use samtools \n",
    "\n",
    "```\n",
    "samtools flagstat <path to bams directory>/input-esc.chr17.2m.bam\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you want to understand better the way SAM files are organized you can follow <a href=\"https://samtools.github.io/hts-specs/SAMv1.pdf\" target=\"_blank\">https://samtools.github.io/hts-specs/SAMv1.pdf</a> section 1.4.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "--- \n",
    "### **2.** Create a “tag directory” \n",
    "\n",
    "For the example Oct4 ChIP-seq experiment using the makeTagDirectory command. Start by typing makeTagDirectory (without any options) in your command line, it will provide the usage, some info about the command and a full list of program options – as I mentioned above, I highly recommend doing that whenever you use a new tool and a new command. \n",
    "\n",
    "Tag directories are analogous to sorted bam files and are the starting point for most HOMER operations like finding peaks, creating visualization files, or calculating read densities. The command also performs several quality control and parameter estimation calculations. The command has the following form:  \n",
    "    \n",
    "```\n",
    "makeTagDirectory <Output Tag Directory> [options] <input SAM/BAM file1> [input SAM/BAM file2] ...\n",
    "``` \n",
    "\n",
    "To create a tag directory for the Oct4 experiment, run the following command with recommended options:\n",
    "\n",
    "```\n",
    "makeTagDirectory <path to where your tag directories will be>/oct4-esc -genome mm9 -checkGC <path to your bam files>/oct4-esc.chr17.2m.bam\n",
    "```\n",
    "\n",
    "The command will take several seconds to run. What it is doing is parsing through the BAM file, removing reads that do not align to a unique position in the genome, separating reads by chromosome and sorting them by position, calculating how often reads appear in the same position to estimate the clonality (i.e. PCR duplication), calculating the relative distribution of reads relative to one another to estimate the ChIP-fragment length, calculating sequence properties and GC-content of the reads and performs a simple enrichment calculation to check if the experiment looks like a ChIP-seq experiment (vs. an RNA-seq experiment).\n",
    "\n",
    "The command creates a new directory, in this case named **oct4-esc**. Inside the directory are several text files that contain various QC results. \n",
    "\n",
    "Try looking at the following by typing:\n",
    "\n",
    "```\n",
    "less -S <filename>\n",
    "```\n",
    "\n",
    "or for all files use:   \n",
    "\n",
    "```\n",
    "less -S tagdirs/oct4-esc/*txt\n",
    "\n",
    "``` \n",
    "\n",
    "and scroll between files with `:n` for next and `:p` for previous.\n",
    "\n",
    "> * **tagInfo.txt** - summary information from the experiment, including read totals.\n",
    "> * **tagFreqUniq.txt** - nucleotide frequencies relative to the 5’ end of the sequencing reads.\n",
    "> * **genomeGCcontent.txt** - distribution of ChIP-fragment GC%\n",
    "> * **tagAutocorrelation.txt** - relative distribution of reads found on the same strand vs. different strands.\n",
    "> * **tagCountDistribution.txt** - number of reads appearing at the same positions.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "--- \n",
    "### **3.** Create “tag directories” for all samples\n",
    "\n",
    "By following using a ‘for loop’. \n",
    "\n",
    "```\n",
    "for f in <path to your bam files>/*.bam; \n",
    "do fname=`basename $f .chr17.2m.bam`; \n",
    "    makeTagDirectory <path to where your tag directories will be>/$fname -genome mm9 -checkGC $f; \n",
    "done\n",
    "```\n",
    "\n",
    "So in my case I ran it like:\n",
    "\n",
    "```\n",
    "for f in ~/goren-module5/bams/*bam;\n",
    "do fname=`basename $f .chr17.2m.bam`; \n",
    "    makeTagDirectory ~/goren-module5/tagdirs/$fname -genome mm9 -checkGC $f; \n",
    "done\n",
    "```\n",
    "\n",
    "At this point you should have 6 tag directories. Look through the QC stats of the various ones."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "### **4.** Next we will visualize the ChIP-seq experiments.\n",
    "\n",
    "By creating bedGraph files from the tag directories and using the IGV genome browser to look at the results. We will do this using the makeUCSCfile command. For most ChIP-seq experiments all you need to do is specify the tag directory and specify “-o auto” for the command to automatically save the bedGraph file inside the tag directory:\n",
    "\n",
    "```\n",
    "makeUCSCfile <Tag Directory> -o auto\n",
    "```\n",
    "\n",
    "For a specific dataset, e.g. Oct4, the command would be:\n",
    "\n",
    "```\n",
    "makeUCSCfile <path to your tag directories>/oct4-esc/ -o auto\n",
    "```\n",
    "\n",
    "This creates the file “oct4-esc/oct4-esc.ucsc.bedGraph.gz”. This file format specifies the normalized read depth at variable intervals along the genome (use zmore and the filename to view the file format for yourself). \n",
    "\n",
    "Now make these for all samples:\n",
    "\n",
    "```\n",
    "for dir in <path to tag directories>/*; \n",
    "do makeUCSCfile $dir -o auto; \n",
    "done\n",
    "```\n",
    "\n",
    "In my case I ran like:\n",
    "```\n",
    "for dir in ~/goren-module5/tagdirs/*; \n",
    "do makeUCSCfile $dir -o auto ; \n",
    "done\n",
    "```\n",
    "\n",
    "\n",
    "### To view the file in the genome browser, do the following:\n",
    "\n",
    "Download the files to your computer \n",
    "\n",
    "``` \n",
    "scp ucsd-train<your number>@tscc.sdsc.edu:/home/ucsd-train<your number>/<full path to the file> <path to location to be copied to> \n",
    "```\n",
    "\n",
    "or for all files:\n",
    "    \n",
    "```\n",
    "scp -r ucsd-train<your number>@tscc.sdsc.edu:/home/ucsd-train<your number>/<full path to the directory>*/*.gz <path to location to be copied to>\n",
    "```\n",
    "\n",
    "In my environment it is:\n",
    "\n",
    "```\n",
    "scp -r ucsd-train01@tscc.sdsc.edu:/home/ucsd-train01/goren-module5/tagdirs/*/*.gz ~/Google\\ Drive//000-all-directories/0-GorenLab/Teaching/BIOM262/2020-biom262-cmm262/module5/data-import/\n",
    "```\n",
    "\n",
    "You can also use FileZilla for this if you have set up a connection with the cluster.\n",
    "\n",
    "\n",
    "\n",
    "**Open IGV.** Make sure you use the right genome (mm9; it is a good practice to have!) and drag the file to the center window (or select file -> load from file).\n",
    "\n",
    "The read pileups will display the relative density of ChIP-seq reads at each position in the genome. We only have data for chr17 in this example, so we can stick to that chromosome.\n",
    "\n",
    "---\n",
    "### **5.** See if there are any interesting patterns in the data that catch your eye.\n",
    "\n",
    "Try visiting the Pou5f1 locus (the gene for Oct4) by typing the gene name into the search bar at the top. Once at the Pou5f1 locus, zoom out (alt+click or scale on top right) to see if there any nearby sites that might resemble enhancers.\n",
    "\n",
    "Each dataset was created by a different antibody, and they can be divided into three types: TFs, HMs and global input. Since we will need to treat each type differently, I recommend making a directory for each – input, tfs and hms and move the tag directories to the relevant one (e.g. tfs/oct4-esc/, etc). \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
